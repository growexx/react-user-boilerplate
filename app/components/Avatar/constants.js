// component name as ga label
export const GA_LABEL_AVATAR = 'Avatar';
