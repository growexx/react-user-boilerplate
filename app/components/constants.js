export const LAYOUT_CONFIG = {
  VERTICAL_OPTION_1: 1,
  VERTICAL_OPTION_2: 2,

  HORIZONTAL_OPTION_1: 3,
};

export const showLogoInSideBar = currentVariant =>
  ![
    LAYOUT_CONFIG.VERTICAL_OPTION_2,
    LAYOUT_CONFIG.HORIZONTAL_OPTION_1,
  ].includes(currentVariant);
