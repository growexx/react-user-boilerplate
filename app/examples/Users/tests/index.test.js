/**
 *
 * Tests for Users
 *
 */

import React from 'react';
import { fireEvent, render, waitForElement } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { browserHistory } from 'react-router-dom';
import { Provider } from 'react-redux';
import history from 'utils/history';
import request from 'utils/request';
import { ConnectedRouter } from 'connected-react-router';
import Users, { mapDispatchToProps } from '../index';
import configureStore from '../../../configureStore';
import {
  addNewUserFailure,
  addNewUserSuccess,
  failedResponse,
  responseWithList,
  responseWithZeroList,
  USER_DATA,
} from '../stub';
import { TEST_IDS } from '../constants';
jest.mock('utils/request');

let store;
const props = {
  pristine: true,
  reset: true,
  submitting: true,
};

const fieldUpdateViaPlaceHolder = [
  {
    key: 'john.doe@growexx.com',
    value: USER_DATA.EMAIL,
  },
  {
    key: 'John',
    value: USER_DATA.NAME,
  },
  {
    key: 'Doe',
    value: USER_DATA.NAME,
  },
];

const componentWrapper = updatedProps =>
  render(
    <Provider store={store}>
      <IntlProvider locale="en">
        <ConnectedRouter history={history}>
          <Users {...props} {...updatedProps} />
        </ConnectedRouter>
      </IntlProvider>
    </Provider>,
  );

describe('Check component:<Users /> is rendering properly', () => {
  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  beforeEach(() => {
    request.mockImplementation(() => Promise.resolve(responseWithZeroList()));
  });

  afterEach(() => {
    request.mockClear();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = componentWrapper();
    expect(firstChild).toMatchSnapshot();
  });

  it('mapDispatch to props', () => {
    const mockFn = jest.fn();
    const returnValue = mapDispatchToProps(mockFn);
    returnValue.updateField(
      { target: { name: 'email' } },
      { target: { value: USER_DATA.EMAIL } },
    );

    expect(mockFn).toBeCalled();
  });

  it('Click: New User Modal should show modal', async () => {
    const { getByTestId, getByText } = componentWrapper();
    // Fire Event
    fireEvent.click(getByTestId(TEST_IDS.ADD_USER));

    // Check Elements are showing
    expect(getByText('Add User')).toBeTruthy();

    fireEvent.click(getByTestId(TEST_IDS.ADD_USER));
    fireEvent.click(getByTestId(TEST_IDS.USER_MODAL_CANCEL));
  });

  it('Click Delete: Show Confirmation Modal', async () => {
    const { getByTestId, getByText } = componentWrapper();
    await waitForElement(() => getByText('Active'));

    // Click Delete Button
    fireEvent.click(getByTestId(TEST_IDS.DELETE_BUTTON));

    // Check Elements are showing
    expect(getByText('OK', { trim: true })).toBeTruthy();
  });

  it('Click Delete: Show Confirmation Modal and click confirm', async () => {
    const { getByTestId, getByText } = componentWrapper();
    await waitForElement(() => getByText('Active'));

    // Click Delete Button
    fireEvent.click(getByTestId(TEST_IDS.DELETE_BUTTON));

    // Check Elements are showing
    expect(getByText('OK', { trim: true })).toBeTruthy();
    fireEvent.click(getByTestId(TEST_IDS.DELETE_BUTTON_CONFIRMED));
  });
});

describe('Check listing of users is rendering properly', () => {
  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  beforeEach(() => {
    request.mockImplementation(() => Promise.resolve(responseWithZeroList()));
  });

  afterEach(() => {
    request.mockClear();
  });

  it('No Records found for users', async () => {
    request.mockImplementation(() => Promise.resolve(responseWithZeroList()));
    const { getByText } = componentWrapper({ demo: false });
    await waitForElement(() => getByText('No Data'));
    expect(getByText('No Data')).toBeTruthy();
  });

  it('Users Listing with few records should be shown', async () => {
    request.mockImplementation(() => Promise.resolve(responseWithList()));
    const { getByText } = componentWrapper({ demo: false });
    await waitForElement(() => getByText('Active'));

    expect(getByText('Active')).toBeTruthy();
  });

  it('Users Listing with pagination', async () => {
    request.mockImplementation(() => Promise.resolve(responseWithList()));
    const { getByText, getByTitle } = componentWrapper({ demo: false });
    await waitForElement(() => getByText('Active'));

    // Expand Data
    fireEvent.click(document.querySelectorAll('.ant-table-row-expand-icon')[0]);

    expect(getByText('Active')).toBeTruthy();
    fireEvent.click(getByTitle('2'));
  });

  it('Failed Users Listing api', async () => {
    request.mockImplementationOnce(() => Promise.reject(failedResponse));
    const { getByText } = componentWrapper({ demo: false });
    await waitForElement(() => getByText('No Data'));
    expect(getByText('No Data')).toBeTruthy();
  });

  it('Toggle User Status', async () => {
    request.mockImplementation(() => Promise.resolve(responseWithList()));

    const { getByTestId, getByText } = componentWrapper({
      demo: false,
    });
    // Wait till data shows
    await waitForElement(() => getByText('Active'));

    // Fire Event
    fireEvent.click(getByTestId(TEST_IDS.STATUS_TOGGLE));
  });

  it('Click Delete: Show Confirmation Modal and click confirm', async () => {
    request.mockImplementation(() => Promise.resolve(failedResponse()));

    const { getByTestId, getByText } = componentWrapper();
    await waitForElement(() => getByText('Active'));

    // Click Delete Button
    fireEvent.click(getByTestId(TEST_IDS.DELETE_BUTTON));

    // Check Elements are showing
    expect(getByText('OK', { trim: true })).toBeTruthy();
    fireEvent.click(getByTestId(TEST_IDS.DELETE_BUTTON_CONFIRMED));
  });

  it('Click Delete: Show Confirmation Modal and click confirm', async () => {
    request.mockImplementation(() => Promise.resolve(failedResponse()));

    const { getByTestId, getByText } = componentWrapper();
    await waitForElement(() => getByText('Active'));

    // Click Delete Button
    fireEvent.click(getByTestId(TEST_IDS.DELETE_BUTTON));

    // Check Elements are showing
    expect(getByText('OK', { trim: true })).toBeTruthy();
    fireEvent.click(getByTestId(TEST_IDS.DELETE_CONFIRMATION_CANCEL));
  });

  it('Toggle User Status Local', async () => {
    request.mockImplementation(() => Promise.resolve(responseWithList()));

    const { getByTestId, getByText } = componentWrapper({
      demo: true,
    });
    // Wait till data shows
    await waitForElement(() => getByText('Active'));

    // Fire Event
    fireEvent.click(getByTestId(TEST_IDS.STATUS_TOGGLE));
  });

  it('Toggle User Status Local', async () => {
    request.mockImplementation(() => Promise.resolve(failedResponse()));

    const { getByTestId, getByText } = componentWrapper({
      demo: true,
    });
    // Wait till data shows
    await waitForElement(() => getByText('Active'));

    // Fire Event
    fireEvent.click(getByTestId(TEST_IDS.STATUS_TOGGLE));
  });
});

describe('New Users', () => {
  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  beforeEach(() => {
    request.mockImplementation(() => Promise.resolve(responseWithList()));
  });

  afterEach(() => {
    request.mockClear();
  });

  it('Add new users with success', () => {
    request.mockImplementation(() => Promise.resolve(addNewUserSuccess()));

    const { getByTestId, getByPlaceholderText, getByText } = componentWrapper({
      demo: false,
    });
    // Fire Event
    fireEvent.click(getByTestId(TEST_IDS.ADD_USER));

    // Update Fields
    fieldUpdateViaPlaceHolder.forEach(d => {
      fireEvent.change(getByPlaceholderText(d.key), {
        target: { value: d.value },
      });
    });
    // Check Elements are showing
    expect(getByText('Add User')).toBeTruthy();
    fireEvent.click(getByText('Add'));
  });

  it('Add new users with success', () => {
    const { getByTestId, getByPlaceholderText, getByText } = componentWrapper({
      demo: true,
    });
    // Fire Event
    fireEvent.click(getByTestId(TEST_IDS.ADD_USER));

    // Update Fields
    fieldUpdateViaPlaceHolder.forEach(d => {
      fireEvent.change(getByPlaceholderText(d.key), {
        target: { value: d.value },
      });
    });
    // Check Elements are showing
    expect(getByText('Add User')).toBeTruthy();
    fireEvent.click(getByText('Add'));
  });

  it('Add new user with cancel option', () => {
    request.mockImplementation(() => Promise.resolve(addNewUserSuccess()));

    const { getByTestId } = componentWrapper({ demo: false });
    // Fire Event
    fireEvent.click(getByTestId(TEST_IDS.ADD_USER));

    fireEvent.click(getByTestId(TEST_IDS.USER_MODAL_CANCEL));
  });

  it('Add new user with failure', () => {
    request.mockImplementation(() => Promise.reject(addNewUserFailure()));

    const { getByTestId, getByText, getByPlaceholderText } = componentWrapper({
      demo: false,
    });
    // Fire Event
    fireEvent.click(getByTestId(TEST_IDS.ADD_USER));

    // Update Fields
    fieldUpdateViaPlaceHolder.forEach(d => {
      fireEvent.change(getByPlaceholderText(d.key), {
        target: { value: d.value },
      });
    });

    // Check Elements are showing
    expect(getByText('Add User')).toBeTruthy();
    fireEvent.click(getByTestId(TEST_IDS.USER_MODAL_OK));
  });
});

describe('Update User', () => {
  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  beforeEach(() => {
    request.mockImplementation(() => Promise.resolve(responseWithList()));
  });

  afterEach(() => {
    request.mockClear();
  });

  it('Update user with success', async () => {
    request.mockImplementationOnce(() => Promise.resolve(responseWithList()));

    const { getByTestId, getByText, getByPlaceholderText } = componentWrapper({
      demo: false,
    });
    expect(request).toHaveBeenCalledTimes(1);
    await waitForElement(() => getByText('Active'));

    // Fire Event
    fireEvent.click(getByTestId(TEST_IDS.EDIT_BUTTON));

    // Update Fields
    fieldUpdateViaPlaceHolder.forEach(d => {
      fireEvent.change(getByPlaceholderText(d.key), {
        target: { value: d.value },
      });
    });

    // Check Elements are showing
    expect(getByText('Update')).toBeTruthy();
    fireEvent.click(getByTestId(TEST_IDS.ADD_USER));
  });

  it('Update user with success', async () => {
    request.mockImplementationOnce(() => Promise.resolve(responseWithList()));

    const { getByTestId, getByText, getByPlaceholderText } = componentWrapper({
      demo: true,
    });
    await waitForElement(() => getByText('Active'));

    // Fire Event
    fireEvent.click(getByTestId(TEST_IDS.EDIT_BUTTON));

    // Update Fields
    fieldUpdateViaPlaceHolder.forEach(d => {
      fireEvent.change(getByPlaceholderText(d.key), {
        target: { value: d.value },
      });
    });

    // Check Elements are showing
    expect(getByText('Update')).toBeTruthy();
    fireEvent.click(getByText('Update'));
  });

  it('Update user with cancel', async () => {
    request.mockImplementationOnce(() => Promise.resolve(responseWithList()));

    const { getByTestId, getByText } = componentWrapper({ demo: false });
    expect(request).toHaveBeenCalledTimes(1);
    await waitForElement(() => getByText('Active'));

    // Fire Event
    fireEvent.click(getByTestId(TEST_IDS.EDIT_BUTTON));

    // Check Elements are showing
    expect(getByText('Update')).toBeTruthy();
    fireEvent.click(getByTestId(TEST_IDS.USER_MODAL_CANCEL));
  });

  it('User Update failure', async () => {
    request
      .mockImplementationOnce(() => Promise.resolve(responseWithList()))
      .mockImplementationOnce(() => Promise.resolve(addNewUserFailure()));

    const { getByTestId, getByText, getByPlaceholderText } = componentWrapper({
      demo: false,
    });
    expect(request).toHaveBeenCalledTimes(1);
    await waitForElement(() => getByText('Active') || getByText('Suspended'));

    // Fire Event
    fireEvent.click(getByTestId(TEST_IDS.EDIT_BUTTON));

    // Update Fields
    fieldUpdateViaPlaceHolder.forEach(d => {
      fireEvent.change(getByPlaceholderText(d.key), {
        target: { value: d.value },
      });
    });

    // Check Elements are showing
    expect(getByText('Update')).toBeTruthy();
    fireEvent.click(getByText('Update'));
  });
});

describe('Status Filter', () => {
  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  beforeEach(() => {
    request.mockImplementation(() => Promise.resolve(responseWithList()));
  });

  afterEach(() => {
    request.mockClear();
  });

  it('Status Filter', async () => {
    request.mockImplementationOnce(() => Promise.resolve(responseWithList()));

    const { getByText, getByRole } = componentWrapper({
      demo: false,
    });
    await waitForElement(() => getByRole('combobox'));

    fireEvent.mouseDown(getByRole('combobox'));
    fireEvent.change(getByRole('combobox'), {
      target: {
        value: 'Active',
      },
    });
    // .ant-select-item-option-content
    fireEvent.click(
      document.querySelectorAll('.ant-select-item-option-content')[0],
    );
    fireEvent.blur(getByRole('combobox'));
    fireEvent.focus(getByRole('combobox'));

    // Update Fields
    fireEvent.click(getByText('Name'));
  });
});

describe('Search & Sorting user list', () => {
  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  beforeEach(() => {
    request.mockImplementation(() => Promise.resolve(responseWithList()));
  });

  afterEach(() => {
    request.mockClear();
  });

  it('Sorting user with success', async () => {
    request.mockImplementationOnce(() => Promise.resolve(responseWithList()));

    const { getByText } = componentWrapper({
      demo: false,
    });

    // Update Fields
    fireEvent.click(getByText('Name'));
  });

  it('Search user with success', async () => {
    request.mockImplementationOnce(() => Promise.resolve(responseWithList()));

    componentWrapper({
      demo: false,
    });

    document.querySelectorAll('.ant-input-affix-wrapper')[0].children[0].value =
      'Hello';

    fireEvent.click(document.querySelectorAll('.ant-input-search-button')[0]);
  });

  it('Search user with success', async () => {
    const { getByPlaceholderText } = componentWrapper({});

    // Update Fields
    fireEvent.change(getByPlaceholderText('Search User'), {
      target: { value: 'a' },
    });

    // Update Fields
    fireEvent.change(getByPlaceholderText('Search User'), {
      target: { value: '' },
    });
  });

  it('Sorting', async () => {
    const { getByText } = componentWrapper({});

    // Update Fields
    fireEvent.click(getByText('Name'));
    fireEvent.click(getByText('Name'));
    fireEvent.click(getByText('Name'));
    fireEvent.click(getByText('User Id'));
    fireEvent.click(getByText('User Id'));
    fireEvent.click(getByText('User Id'));
  });
});
