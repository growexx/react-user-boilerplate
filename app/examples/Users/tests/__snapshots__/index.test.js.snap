// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check component:<Users /> is rendering properly Should render and match the snapshot 1`] = `
.c3 {
  width: 100%;
  max-width: 400px;
}

.c4 {
  max-width: 500px;
}

.c0 .ant-page-header-heading-extra {
  width: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  width: 100%;
  vertical-align: middle;
  margin: 1px;
}

.c2 .ant-space-item {
  width: 100%;
}

.c6 {
  min-width: 1000px;
}

.c6 .ant-table-content {
  overflow-x: auto;
  min-width: 1000px;
}

.c6 .ant-table-cell {
  background-color: #fff;
}

.c5 {
  box-sizing: border-box;
  margin: 0;
  color: rgba(0,0,0,0.85);
  font-size: 14px;
  font-variant: tabular-nums;
  line-height: 1.5715;
  list-style: none;
  font-feature-settings: 'tnum';
  position: relative;
  padding: 16px 24px;
  overflow-x: auto;
}

@media all and (min-width:768px) {
  .c0 .ant-page-header-heading-extra {
    width: auto;
  }
}

@media all and (min-width:768px) and (max-width:1024px) {
  .c1 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    margin: 0 5px 0 0;
  }
}

@media all and (min-width:480px) and (max-width:768px) {
  .c1 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media all and (max-width:480px) {
  .c1 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div>
  <div
    class="ant-page-header c0 ant-page-header-ghost"
  >
    <div
      class="ant-page-header-heading"
    >
      <div
        class="ant-page-header-heading-left"
      >
        <span
          class="ant-page-header-heading-title"
          title="Users"
        >
          Users
        </span>
      </div>
      <span
        class="ant-page-header-heading-extra"
      >
        <div
          class="c1"
        >
          <div
            class="ant-space ant-space-horizontal ant-space-align-center c2"
          >
            <div
              class="ant-space-item"
            >
              <div
                class="ant-select c3 ant-select-single ant-select-allow-clear ant-select-show-arrow ant-select-disabled"
              >
                <div
                  class="ant-select-selector"
                >
                  <span
                    class="ant-select-selection-search"
                  >
                    <input
                      aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                      aria-autocomplete="list"
                      aria-controls="rc_select_TEST_OR_SSR_list"
                      aria-expanded="false"
                      aria-haspopup="listbox"
                      aria-owns="rc_select_TEST_OR_SSR_list"
                      autocomplete="off"
                      class="ant-select-selection-search-input"
                      disabled=""
                      id="rc_select_TEST_OR_SSR"
                      readonly=""
                      role="combobox"
                      style="opacity: 0;"
                      type="search"
                      unselectable="on"
                      value=""
                    />
                  </span>
                  <span
                    class="ant-select-selection-placeholder"
                  >
                    User Status
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="ant-select-arrow"
                  style="user-select: none;"
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    class="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
          <div
            class="ant-space ant-space-horizontal ant-space-align-center c2"
          >
            <div
              class="ant-space-item"
            >
              <span
                class="ant-input-group-wrapper ant-input-search c4"
              >
                <span
                  class="ant-input-wrapper ant-input-group"
                >
                  <span
                    class="ant-input-affix-wrapper"
                  >
                    <input
                      class="ant-input"
                      placeholder="Search User"
                      type="text"
                      value=""
                    />
                    <span
                      class="ant-input-suffix"
                    >
                      <span
                        aria-label="close-circle"
                        class="anticon anticon-close-circle ant-input-clear-icon-hidden ant-input-clear-icon"
                        role="button"
                        tabindex="-1"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="close-circle"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                          />
                        </svg>
                      </span>
                    </span>
                  </span>
                  <span
                    class="ant-input-group-addon"
                  >
                    <button
                      class="ant-btn ant-btn-icon-only ant-input-search-button"
                      type="button"
                    >
                      <span
                        aria-label="search"
                        class="anticon anticon-search"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="search"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                          />
                        </svg>
                      </span>
                    </button>
                  </span>
                </span>
              </span>
            </div>
          </div>
          <div
            class="ant-space ant-space-horizontal ant-space-align-center c2"
          >
            <div
              class="ant-space-item"
            >
              <button
                class="ant-btn"
                color="primary"
                data-testid="ADD_USER"
                type="button"
              >
                <span>
                  Add User
                </span>
              </button>
            </div>
          </div>
        </div>
      </span>
    </div>
  </div>
  <div
    class="c5"
  >
    <div
      class="ant-table-wrapper c6"
    >
      <div
        class="ant-spin-nested-loading"
      >
        <div>
          <div
            class="ant-spin ant-spin-spinning"
          >
            <span
              class="ant-spin-dot ant-spin-dot-spin"
            >
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
            </span>
          </div>
        </div>
        <div
          class="ant-spin-container ant-spin-blur"
        >
          <div
            class="ant-table ant-table-empty"
          >
            <div
              class="ant-table-container"
            >
              <div
                class="ant-table-content"
              >
                <table
                  style="table-layout: auto;"
                >
                  <colgroup>
                    <col
                      class="ant-table-expand-icon-col"
                    />
                    <col
                      style="width: 10%; min-width: 10%;"
                    />
                    <col
                      style="width: 20%; min-width: 20%;"
                    />
                    <col
                      style="width: 20%; min-width: 20%;"
                    />
                    <col
                      style="width: 15%; min-width: 15%;"
                    />
                    <col
                      style="width: 15%; min-width: 15%;"
                    />
                    <col
                      style="width: 20%; min-width: 20%;"
                    />
                  </colgroup>
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <th
                        class="ant-table-cell ant-table-row-expand-icon-cell"
                      />
                      <th
                        class="ant-table-cell ant-table-column-has-sorters"
                      >
                        <div
                          class="ant-table-column-sorters-with-tooltip"
                        >
                          <div
                            class="ant-table-column-sorters"
                          >
                            <span>
                              User Id
                            </span>
                            <span
                              class="ant-table-column-sorter ant-table-column-sorter-full"
                            >
                              <span
                                class="ant-table-column-sorter-inner"
                              >
                                <span
                                  aria-label="caret-up"
                                  class="anticon anticon-caret-up ant-table-column-sorter-up"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="caret-up"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                    />
                                  </svg>
                                </span>
                                <span
                                  aria-label="caret-down"
                                  class="anticon anticon-caret-down ant-table-column-sorter-down"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="caret-down"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                    />
                                  </svg>
                                </span>
                              </span>
                            </span>
                          </div>
                        </div>
                      </th>
                      <th
                        class="ant-table-cell ant-table-column-has-sorters"
                      >
                        <div
                          class="ant-table-column-sorters-with-tooltip"
                        >
                          <div
                            class="ant-table-column-sorters"
                          >
                            <span>
                              Name
                            </span>
                            <span
                              class="ant-table-column-sorter ant-table-column-sorter-full"
                            >
                              <span
                                class="ant-table-column-sorter-inner"
                              >
                                <span
                                  aria-label="caret-up"
                                  class="anticon anticon-caret-up ant-table-column-sorter-up"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="caret-up"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                    />
                                  </svg>
                                </span>
                                <span
                                  aria-label="caret-down"
                                  class="anticon anticon-caret-down ant-table-column-sorter-down"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="caret-down"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                    />
                                  </svg>
                                </span>
                              </span>
                            </span>
                          </div>
                        </div>
                      </th>
                      <th
                        class="ant-table-cell ant-table-column-has-sorters"
                      >
                        <div
                          class="ant-table-column-sorters-with-tooltip"
                        >
                          <div
                            class="ant-table-column-sorters"
                          >
                            <span>
                              Email
                            </span>
                            <span
                              class="ant-table-column-sorter ant-table-column-sorter-full"
                            >
                              <span
                                class="ant-table-column-sorter-inner"
                              >
                                <span
                                  aria-label="caret-up"
                                  class="anticon anticon-caret-up ant-table-column-sorter-up"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="caret-up"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                    />
                                  </svg>
                                </span>
                                <span
                                  aria-label="caret-down"
                                  class="anticon anticon-caret-down ant-table-column-sorter-down"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="caret-down"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                    />
                                  </svg>
                                </span>
                              </span>
                            </span>
                          </div>
                        </div>
                      </th>
                      <th
                        class="ant-table-cell ant-table-column-has-sorters"
                      >
                        <div
                          class="ant-table-column-sorters-with-tooltip"
                        >
                          <div
                            class="ant-table-column-sorters"
                          >
                            <span>
                              Status
                            </span>
                            <span
                              class="ant-table-column-sorter ant-table-column-sorter-full"
                            >
                              <span
                                class="ant-table-column-sorter-inner"
                              >
                                <span
                                  aria-label="caret-up"
                                  class="anticon anticon-caret-up ant-table-column-sorter-up"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="caret-up"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                    />
                                  </svg>
                                </span>
                                <span
                                  aria-label="caret-down"
                                  class="anticon anticon-caret-down ant-table-column-sorter-down"
                                  role="img"
                                >
                                  <svg
                                    aria-hidden="true"
                                    data-icon="caret-down"
                                    fill="currentColor"
                                    focusable="false"
                                    height="1em"
                                    viewBox="0 0 1024 1024"
                                    width="1em"
                                  >
                                    <path
                                      d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                    />
                                  </svg>
                                </span>
                              </span>
                            </span>
                          </div>
                        </div>
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Last Access Date
                      </th>
                      <th
                        class="ant-table-cell"
                      >
                        Actions
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    class="ant-table-tbody"
                  >
                    <tr
                      class="ant-table-placeholder"
                    >
                      <td
                        class="ant-table-cell"
                        colspan="7"
                      >
                        <div
                          class="ant-empty ant-empty-normal"
                        >
                          <div
                            class="ant-empty-image"
                          >
                            <svg
                              class="ant-empty-img-simple"
                              height="41"
                              viewBox="0 0 64 41"
                              width="64"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <g
                                fill="none"
                                fill-rule="evenodd"
                                transform="translate(0 1)"
                              >
                                <ellipse
                                  class="ant-empty-img-simple-ellipse"
                                  cx="32"
                                  cy="33"
                                  rx="32"
                                  ry="7"
                                />
                                <g
                                  class="ant-empty-img-simple-g"
                                  fill-rule="nonzero"
                                >
                                  <path
                                    d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                  />
                                  <path
                                    class="ant-empty-img-simple-path"
                                    d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                  />
                                </g>
                              </g>
                            </svg>
                          </div>
                          <div
                            class="ant-empty-description"
                          >
                            No Data
                          </div>
                        </div>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
